site/scripts/prepare-blocks.ts

```diff
@@ -0,0 +1,483 @@
+/**
+ * This script converts JSON block infos into a folder with static block files.
+ * It downloads and builds block sources for that. The result of this script is
+ * then served by the Next.js app on Vercel.
+ *
+ * To avoid unnecessary rebuilds, this script puts unstable_hubInfo.checksum into
+ * each block-metadata.json file. This checksum is matched against the block info
+ * in subsequent runs to find cache matches.
+ *
+ * We plan to separate block artifacts from the Vercel deployment in the future.
+ * This script will be removed, but its parts will be recycled.
+ */
+import chalk from "chalk";
+import execa from "execa";
+import path from "path";
+import fs from "fs-extra";
+import glob from "glob";
+import micromatch from "micromatch";
+import * as envalid from "envalid";
+import md5 from "md5";
+import Ajv, { JSONSchemaType } from "ajv";
+import tmp from "tmp-promise";
+import slugify from "slugify";
+
+const monorepoRoot = path.resolve(__dirname, "../..");
+
+const defaultExecaOptions = {
+  env: {
+    NODE_ENV: "development",
+    PATH: process.env.PATH,
+  },
+  extendEnv: false,
+  stdio: "inherit",
+} as const;
+
+interface StoredBlockInfo {
+  repository: string;
+  commit: string;
+
+  distDir?: string;
+  folder?: string;
+  workspace?: string;
+}
```

Comment:
There's an existing type called `BuildConfig` [here](https://github.com/blockprotocol/blockprotocol/blob/335f799a2fb242256e256925b0c3496f5b7c267b/site/src/lib/blocks.ts#L14)  - we should share one type across these files.

You'll see that `readBlocksFromDisk` is [relying](https://github.com/blockprotocol/blockprotocol/blob/335f799a2fb242256e256925b0c3496f5b7c267b/site/src/lib/blocks.ts#L94) on there being a `branch`, leading to things like the below - we'll need to update this. Potentially to include the commit sha in the URL (we could always hide it from the actual displayed link as it'll be pretty ugly - e.g. just make the display text the repo root, with the link actually going to the specific commit & folder).

The `timestamp` is also relied on to generate a 'last updated at' display on the block page (where timestamp exists) - we should at least remove this for now. Obviously ideally we'd get the last updated time from the commit or similar.

![Screenshot 2022-02-23 at 18 43 38](https://user-images.githubusercontent.com/37743469/155387009-cd55741e-3c09-4ca0-a5b3-0b74938222b9.png)

Reply:
Good spot, thanks Ciaran! I merged the two typings in 1b97e39f2cb045d6ec59b65b044fe3cb2f102d25. I kept `lastUpdated` in the UI, but hardcoded it to null for now inside `site/src/lib/blocks.ts`. We can extract commit metadata via GitHub API in another PR and bring it back.

Links seem to work now:

<img width="450" alt="Screenshot 2022-02-25 at 09 59 34" src="https://user-images.githubusercontent.com/608862/155696069-97d239e4-0358-4605-9010-f1250a08e125.png">
<img width="512" alt="Screenshot 2022-02-25 at 09 59 51" src="https://user-images.githubusercontent.com/608862/155696078-fb6c491c-87c4-4963-a14c-467f8eb98f70.png">

Reply:
I'd still be tempted to include the commit sha in the URL - then you can see the source which is actually powering the block on the hub (@thehabbos007's comment). We don't have to worry about versions since we always show one version on the hub, and we know what exact commit that is built from. 

`getRepositoryUrl(`${repositoryUrl}#${commitHash}`)` should give us the link with commit automatically, which we can then truncate for display purposes to only the repo itself (so we don't have horribly long display URLs).

I don't think this is vital, but it seems a cheap thing to do which will make sure people who care to inspect the source will see the source that is actually being used - are there reasons not to?

Reply:
I’ve added commit hashes to links. So for example Callout block refers to 

https://github.com/hashintel/hash/tree/ec9e50601629207f1fde76a817b02aa8733762fe/packages/blocks/callout

This is how it is rendered:
<img width="552" alt="Screenshot 2022-03-03 at 15 16 05" src="https://user-images.githubusercontent.com/608862/156582590-2fca53db-918e-49b0-94b9-82c9f4e1bf25.png">

Not ideal, but might work for now. It might be better to have something like this in the future:

> Origin 🔗
>
> :octocat: hashintel/hash
> #️⃣ ec9e50601629207f1fde76a817b02aa8733762fe
> 📁 packages/blocks/callout

↑ That’s assuming that we still hard-wire our hub to git repos instead of having a publish process where build files end up in our storage.